### PR TITLE
Fix unmount state warning.

### PR DESCRIPTION
**Pull request by danielgent**
## Summary

Addresses issue #3226

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by danielgent** [comment]
Tests :heavy_check_mark: 

As discussed, just `E2E Tests / E2E (WordPress 4.9, Gutenberg 4.9` is failing

**Comment by danielgent** [comment]
**BEFORE**

before


**AFTER**

after


This ticket explains the problem
https://github.com/google/site-kit-wp/issues/3321

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 1]
```diff
@@ -824,8 +824,9 @@ const baseSelectors = {
 		invariant( moduleSlug, 'slug is required.' );
 		const modules = select( STORE_NAME ).getModules();
 
-		if ( ! modules ) {
-			return null;
+		// Return `undefined` if modules haven't been loaded yet.
+		if ( modules === undefined ) {
+			return undefined;
```
Comment (danielgent) [inline assets/js/googlesitekit/modules/datastore/modules.js:829]: As per Evan's comment https://github.com/google/site-kit-wp/issues/3226#issuecomment-834628021

Just to note, this file has several different ways of checking modules for being `null` or `undefined` or falsey, and then returns different things :see_no_evil:

Comment (eugene-manuilov) [inline assets/js/googlesitekit/modules/datastore/modules.js:829]: @danielgent could you, please, also update jsdoc for this selector to mentioned that it can return an undefined?

File: assets/js/components/module/ModuleApp.js [hunk 2]
```diff
@@ -68,8 +68,8 @@ function ModuleApp( { moduleSlug } ) {
 				{ moduleConnected && <DateRangeSelector /> }
 			</Header>
 			<Alert module={ moduleSlug } />
-			{ shouldRenderWidget
-				? (
+			{ shouldRenderWidget && screenWidgetContext &&
```
Comment (eugene-manuilov) [inline assets/js/components/module/ModuleApp.js:71]: Perhaps we need to display a progress bar when `shouldRenderWidget` equals `true` and `screenWidgetContext` is not loaded yet (equals to `undefined`)? Plus, we may probably need to display an error if `shouldRenderWidget` is true again, but `screenWidgetContext` is null because otherwise, users will see just a blank screen.

@aaemnnosttv @felixarntz what do you think?

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @danielgent. Left one comment for you and one for Evan and/or Felix.

File: assets/js/components/module/ModuleApp.js (hunk 2 above)
Comment (danielgent) [reply assets/js/components/module/ModuleApp.js:75]: @aaemnnosttv @felixarntz I'll set this back to code review. The question from Eugene is above though if you want to check it :smile:

File: assets/js/components/module/ModuleApp.js [hunk 3]
```diff
@@ -43,14 +43,15 @@ import HelpMenu from '../help/HelpMenu';
...
+import ProgressBar from '../ProgressBar';
 
 const { useSelect } = Data;
 
 function ModuleApp( { moduleSlug } ) {
 	const helpVisibilityEnabled = useFeature( 'helpVisibility' );
 	const screenWidgetContext = useSelect( ( select ) => select( CORE_MODULES ).getScreenWidgetContext( moduleSlug ) );
 	const moduleConnected = useSelect( ( select ) => select( CORE_MODULES ).isModuleConnected( moduleSlug ) );
-	const shouldRenderWidget = useFeature( 'widgets.moduleScreens' ) && screenWidgetContext;
+	const shouldRenderWidget = useFeature( 'widgets.moduleScreens' );
```
Comment (aaemnnosttv) [inline assets/js/components/module/ModuleApp.js:54]: This is a bit of a convention with `useFeature` – if you look at other uses of the hook, you'll see the variable is `const {feature}Enabled = useFeature(...)` so it would be great to preserve that for consistency.

File: assets/js/components/module/ModuleApp.js (hunk 2 above)
Comment (aaemnnosttv) [reply assets/js/components/module/ModuleApp.js:75]: I think it's a good idea to consider, but probably a little out of scope for this issue. I'm also not sure it would be possible to see this in the current module app since we preload the response that is fetched by `getModules` on page load and this will always be requested immediately on module dashboards so we would never run into a situation where the user would actually wait for any substantial amount of time here.

I think the better solution for now is to render the `WidgetContextRenderer` unconditionally if the feature is enabled and handle the `null` value within that component if necessary. We shouldn't need to worry about `ModuleApp` rendering without a screen context as this will be a requirement once the widgets API is the primary layout mechanism.

TL;DR – let's remove the progress bar for now and not use `screenWidgetContext` as a condition here if we can, and handle that in the context renderer (if needed).

File: assets/js/googlesitekit/modules/datastore/modules.js (hunk 1 above)
Comment (aaemnnosttv) [reply assets/js/googlesitekit/modules/datastore/modules.js:830]: `@return` still needs `undefined` added as a potential type.

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 4]
```diff
@@ -811,6 +811,7 @@ const baseSelectors = {
 	/**
 	 * Gets the module's screenWidgetContext.
 	 *
+	 * Returns `undefined` if dependencies are loading.
 	 * Returns `null` if there is no registered context string for the given module.
 	 * Returns `string` the registered context string, screenWidgetContext for the given module.
```
Comment (aaemnnosttv) [inline assets/js/googlesitekit/modules/datastore/modules.js:816]: Not sure why these are up here 🤔  can we move these down to be part of the `@return` tag instead? This is what we usually do 🙂  (e.g. `getCurrentEntityURL`)

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @danielgent – this looks very close, just a few comments to address. I also responded to @eugene-manuilov 's question below regarding the progress bar.

File: assets/js/components/module/ModuleApp.js (hunk 3 above)
Comment (danielgent) [reply assets/js/components/module/ModuleApp.js]: @aaemnnosttv That sounds really obvious now you've pointed it out! :+1:  

Changed.  I'm using feature flags on other tasks as well so I'll make sure I make the same changes

File: assets/js/googlesitekit/widgets/components/WidgetContextRenderer.js [hunk 5]
```diff
@@ -35,9 +35,17 @@ const { useSelect } = Data;
 const WidgetContextRenderer = ( props ) => {
 	const { slug, className, Header, Footer } = props;
 
-	const widgetAreas = useSelect( ( select ) => select( STORE_NAME ).getWidgetAreas( slug ) );
+	const widgetAreas = useSelect( ( select ) => {
+		if ( slug ) {
+			return	select( STORE_NAME ).getWidgetAreas( slug );
```
Comment (aaemnnosttv) [inline assets/js/googlesitekit/widgets/components/WidgetContextRenderer.js:40]: You have a stray tab in here 😄 
```suggestion
			return select( STORE_NAME ).getWidgetAreas( slug );
```

File: assets/js/googlesitekit/widgets/components/WidgetContextRenderer.js [hunk 6]
```diff
@@ -35,9 +35,17 @@ const { useSelect } = Data;
...
+		}
+	} );
 	const widgetAreasLength = widgetAreas?.length;
 
+	if ( ! slug ) {
```
Comment (aaemnnosttv) [inline assets/js/googlesitekit/widgets/components/WidgetContextRenderer.js:45]: I was thinking maybe we should use `widgetAreasLength` here instead since this is a bit more clear as to why we're returning early but that would be a change that we didn't have before. `slug` was required before and now it isn't, but a context could also have zero areas returned by the selector for the given slug, in which case we probably shouldn't render it since a context with no areas doesn't make much sense I think. Let's do that – we can use `widgetAreas?.length` inline here and below and do away with the extra variable, although the reference beyond this point wouldn't need the conditional for accessing `length` 🙂 

```suggestion
	if ( ! widgetAreas?.length ) {
```

File: assets/js/googlesitekit/widgets/components/WidgetContextRenderer.js [hunk 7]
```diff
@@ -35,9 +35,17 @@ const { useSelect } = Data;
 const WidgetContextRenderer = ( props ) => {
 	const { slug, className, Header, Footer } = props;
 
-	const widgetAreas = useSelect( ( select ) => select( STORE_NAME ).getWidgetAreas( slug ) );
+	const widgetAreas = useSelect( ( select ) => {
+		if ( slug ) {
+			return	select( STORE_NAME ).getWidgetAreas( slug );
+		}
+	} );
```
Comment (aaemnnosttv) [inline assets/js/googlesitekit/widgets/components/WidgetContextRenderer.js:42]: Let's return `null` if there is no `slug` rather than the implicit `undefined`.
```suggestion
		}
		return null;
	} );
```

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @danielgent – this is super close and looking good, just a few final comments on the context renderer and then we'll get this merged. Thanks for your patience!

**Review by aaemnnosttv — APPROVED** [review]
Great, thanks @danielgent !
